{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Análise da sequência e das features presentes no NCBI\n",
    "\n",
    "Começamos por aceder ao NCBI para fazer download de um ficheiro no formato __genbank__ com informação correspondente à nossa zona do genoma, da posição 270001 a 505535.\n",
    "\n",
    "O primeiro passo foi descobrir qual o __genbank_id__ associado ao accession __NC_002942.5__.\n",
    "Para isto usamos a função `Entrez.search`. \n",
    "\n",
    "Após saber o __genbank_id__, obtivemos o __genbank__ usando a função `Entrez.efetch`.\n",
    "\n",
    "```python\n",
    "def fetch_genbank(start, end):\n",
    "    \"\"\"\n",
    "    Procura no NCBI pelo Accession NC_002942.5 e \n",
    "    faz download de toda a informação em formato genbank,\n",
    "    retornando o respectivo record com as features\n",
    "    entre a posição start e end.\n",
    "    \"\"\"\n",
    "\n",
    "    Entrez.email = \"*@*.*\"\n",
    "\n",
    "    handle = Entrez.esearch(db=\"nucleotide\", term=\"NC_002942.5\")\n",
    "    record = Entrez.read(handle)\n",
    "    genbank_id = record[\"IdList\"][0]\n",
    "    handle.close()\n",
    "\n",
    "    handle = Entrez.efetch(db=\"nucleotide\",\n",
    "                           rettype=\"gbwithparts\",\n",
    "                           retmod=\"text\",\n",
    "                           id=genbank_id,\n",
    "                           seq_start=start,\n",
    "                           seq_stop=end)\n",
    "    record = SeqIO.read(handle, \"gb\")\n",
    "    handle.close()\n",
    "\n",
    "    return record\n",
    "```\n",
    "\n",
    "Esta função foi adicionada ao módulo `util.www`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Deste `record` extraímos as features __gene__, __CDS__, __tRNA__ e __rRNA__.\n",
    "\n",
    "```python\n",
    "def extract_features(record):\n",
    "    \"\"\"\n",
    "    Das várias features encontradas no ficheiro genbank, extraímos:\n",
    "      - gene\n",
    "      - CDS\n",
    "      - tRNA\n",
    "      - rRNA\n",
    "\n",
    "    Esta função retorna uma lista apenas com as features\n",
    "    dos tipos indicados acima.\n",
    "    \"\"\"\n",
    "\n",
    "    result = []\n",
    "    features = record.features\n",
    "    feature_count = len(features)\n",
    "    types = [\"gene\", \"CDS\", \"tRNA\", \"rRNA\"]\n",
    "\n",
    "    for i in range(feature_count):\n",
    "        feature = features[i]\n",
    "        should_extract = feature.type in types\n",
    "\n",
    "        if should_extract:\n",
    "            result.append(feature)\n",
    "\n",
    "    return result\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Para finalizar, convertemos esta lista de `record` para um dicionário.\n",
    "A cada um dos `record` retirámos:\n",
    "- db_xref\n",
    "- EC_number\n",
    "- function\n",
    "- gene\n",
    "- gene_synonym\n",
    "- locus_tag\n",
    "- note\n",
    "- product\n",
    "- protein_id\n",
    "- translation\n",
    "\n",
    "Além disso, também ficamos com informação sobre a localização de cada gene e em qual das cadeias está presente.\n",
    "\n",
    "```python\n",
    "def features_to_dictionary(start, features):\n",
    "    \"\"\"\n",
    "    A cada uma das features, extraimos a seguintes propriedades:\n",
    "      - db_xref\n",
    "      - EC_number\n",
    "      - function\n",
    "      - gene\n",
    "      - gene_synonym\n",
    "      - locus_tag\n",
    "      - note\n",
    "      - product\n",
    "      - protein_id\n",
    "      - translation\n",
    "\n",
    "    Também extraímos a localização:\n",
    "      - start\n",
    "      - end\n",
    "      - strand\n",
    "    \"\"\"\n",
    "    dictionary = {}\n",
    "    properties = [\"db_xref\",\n",
    "                  \"EC_number\",\n",
    "                  \"function\",\n",
    "                  \"gene\",\n",
    "                  \"gene_synonym\",\n",
    "                  \"note\",\n",
    "                  \"product\",\n",
    "                  \"protein_id\",\n",
    "                  \"translation\"]\n",
    "\n",
    "    for feature in features:\n",
    "        tag = feature.qualifiers[\"locus_tag\"][0]\n",
    "\n",
    "        if not tag in dictionary:\n",
    "            # se esta tag ainda não existe,\n",
    "            # criar um dicionário vazio para ela\n",
    "            dictionary[tag] = {}\n",
    "\n",
    "        location = feature.location\n",
    "        dictionary[tag][\"location\"] = {}\n",
    "        dictionary[tag][\"location\"][\"start\"] = location._start + start\n",
    "        dictionary[tag][\"location\"][\"end\"] = location._end + start - 1\n",
    "        dictionary[tag][\"location\"][\"strand\"] = location._strand\n",
    "\n",
    "        if feature.type in [\"tRNA\", \"rRNA\"]:\n",
    "            # Taggar as features com estes dois tipos\n",
    "            dictionary[tag][\"type\"] = feature.type\n",
    "        else:\n",
    "            # se não, taggar com \"mRNA\"\n",
    "            dictionary[tag][\"type\"] = \"mRNA\"\n",
    "\n",
    "        for prop in properties:\n",
    "            if prop in feature.qualifiers:\n",
    "                value = feature.qualifiers[prop][0]\n",
    "\n",
    "                if prop == \"db_xref\":\n",
    "                    # renomear esta propriedade e remover \"GeneID:\"\n",
    "                    prop = \"gene_id\"\n",
    "                    value = value[7:]\n",
    "                    \n",
    "                if prop == \"translation\":\n",
    "                    # se translation, também guardar o tamanho\n",
    "                    dictionary[tag][\"length\"] = len(value)\n",
    "\n",
    "                if prop in dictionary[tag]:\n",
    "                    # se já encontramos esta propriedade\n",
    "                    # para este locus_tag\n",
    "                    # verificar que é a mesma\n",
    "                    current_value = dictionary[tag][prop]\n",
    "                    assert value == current_value\n",
    "                else:\n",
    "                    # se não, adicionar\n",
    "                    dictionary[tag][prop] = value\n",
    "\n",
    "    return dictionary\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No final:\n",
    "\n",
    "```python\n",
    "import util.www as www\n",
    "\n",
    "start = 270001\n",
    "end = 505535\n",
    "\n",
    "record = www.fetch_genbank(start, end)\n",
    "features = extract_features(record)\n",
    "dictionary = features_to_dictionary(start, features)\n",
    "print(dictionary)\n",
    "```\n",
    "\n",
    "```json\n",
    "{\n",
    "  \"lpg0232\": {\n",
    "    \"function\": \"Transcription factors / DNA binding proteins\",\n",
    "    \"gene\": \"np20\",\n",
    "    \"gene_id\": \"19831799\",\n",
    "    \"length\": 177,\n",
    "    \"location\": {\n",
    "      \"end\": 270569,\n",
    "      \"start\": 270036,\n",
    "      \"strand\": 1\n",
    "    },\n",
    "    \"product\": \"transcriptional regulator np20\",\n",
    "    \"protein_id\": \"YP_094286.1\",\n",
    "    \"translation\": \"MIGCCLI...\",\n",
    "    \"type\": \"mRNA\"\n",
    "  },\n",
    "  \"lpg0233\": {\n",
    "    \"EC_number\": \"4.1.1.7\",\n",
    "    \"function\": \"Biodegradation of Xenobiotics\",\n",
    "    \"gene\": \"mdlC\",\n",
    "    \"gene_id\": \"19831800\",\n",
    "    \"length\": 530,\n",
    "    \"location\": {\n",
    "      \"end\": 272278,\n",
    "      \"start\": 270686,\n",
    "      \"strand\": -1\n",
    "    },\n",
    "    \"product\": \"benzoylformate decarboxylase\",\n",
    "    \"protein_id\": \"YP_094287.1\",\n",
    "    \"translation\": \"MKKTGSDVLK...\",\n",
    "    \"type\": \"mRNA\"\n",
    "  },\n",
    "  // ...\n",
    "}\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
